singularity-eos/eos/get_sg_eos_functors.hpp

```diff
@@ -216,12 +277,144 @@ struct final_functor {
     if (do_s) {
       sie_v(i) /= mass_sum;
     }
+    check_all_vals(i, npte, tid, pte_converged);
     /* reset mass fractions to original values if not normalized to 1 */
     for (int m = 0; m < nmat; ++m) {
       frac_mass_v(i, m) *= mass_sum;
     }
     return;
   }
+
+ private:
+  // Check to make sure returned values are normal or zero
+  template <typename valT>
+  PORTABLE_FORCEINLINE_FUNCTION int bad_val(valT const value,
+                                            const char *const var_name) const {
+    const bool good =
+        (value == valT{0} || (std::isnormal(1e10 * value) && std::isnormal(value)));
+    if (not good) {
+      using PortsOfCall::printf;
+      printf("### ERROR: Bad singularity-eos output\n  Var: %s\n  Value: %e\n", var_name,
+             value);
+    }
+    return not good;
+  }
+  // Check for prrecluding zero and negative (non-positive) values
+  template <typename valT>
+  PORTABLE_FORCEINLINE_FUNCTION int non_positive_val(valT const value,
+                                                     const char *const var_name) const {
+    const bool not_positive = value <= valT{0};
+    if (not_positive) {
+      using PortsOfCall::printf;
+      printf("### ERROR: Bad singularity-eos output\n  Var: %s\n  Value: %e\n", var_name,
+             value);
```

Comment:
Do we have a `PORTABLE_WARN` method? If so this would be an optimal use-case. Also I think these check functions belong in global scope in a utility header.

Reply:
I'm resistant to using the portable warn for two reasons: 1) this isn't a warning since it's actually impetus for the error and 2) I just didn't want to deal with lifting the string concatenation out of the base class into a utility header.

Now that I need to generate more utility functions though, it's probably worth it to just put the string concatenation in there too. But I'm still resistant to labeling these as warnings.

To illustrate my point, this is a sample output that I think is pretty effective:
```
### ERROR: Bad singularity-eos output
  Var: frac_ie
  Value: 9.656546e-309
### ERROR: Bad singularity-eos output
  Var: temp
  Value: 3.474671e-313
### Bad Value Output state:
  ~~Bulk state~~
    Pressure      :    3.084860764090054e+12
    Temperature   :    3.47467079861110e-313
    Specific IE   :    2.789552895739220e+09
    Bulk Modulus  :    1.127154852753680e+13
    Heat Capacity :    2.937634386418220e+08
    dPdE          :        0.611123943858706
  ~~Material States~~
    mat:            Mass fraction              Material IE          Material Volume
      1:     4.90001520366884e-05   9.656546035209074e-309      0.00053771637633576
      2:        0.999566645882337    7.229168124727389e+08     0.000283501725784361
      4:     0.000384353965626437   -9.152107585263315e+04     6.39064862153302e-06
  ~~PTE iteration *DID NOT* converge~~
### ERROR
  Message:     Bad value returned from singularity-eos interface. See output for details
  File:        /usr/projects/eap/users/jhp/XCAP/singularity-eos/github/singularity-eos/singularity-eos/eos/get_sg_eos_functors.hpp
  Line number: 413
```

I think if the `ERROR` messages said `WARNING`, it might falsely convey that the values aren't the _cause_ of the eventual error.

Reply:
As I was writing this, I realize that this kind of takes the role of an error information function. If what I wrote is useful, it can be moved to `ports_of_call` later.